src/HDF5.jl

```diff
@@ -1593,7 +1593,7 @@ function __init__()
         ENV["HDF5_USE_FILE_LOCKING"] = "FALSE"
     end
 
-    Filters.register_blosc()
+    Filters.register()
```

Comment:
Do we know the performance hit we take by registering additional filters? I suspect it's not large, but we should at least have some idea.

Reply:
Do you mean how this affects the `__init__` time for the module?

Reply:
Released 0.15.7:
```julia
julia> @time using HDF5
  0.299217 seconds (139.98 k allocations: 9.943 MiB, 46.28% compilation time)

(HDF5TempEnv) pkg> st HDF5
      Status `C:\Users\kittisopikulm\HDF5TempEnv\Project.toml`
  [f67ccb44] HDF5 v0.15.7
```

Master:
```julia
julia> @time using HDF5
  0.471673 seconds (162.15 k allocations: 11.446 MiB, 62.28% compilation time)

(HDF5TempEnv) pkg> st HDF5
      Status `C:\Users\kittisopikulm\HDF5TempEnv\Project.toml`
  [f67ccb44] HDF5 v0.16.0 `https://github.com/JuliaIO/HDF5.jl.git#master`
```

After this PR:


```julia
julia> @time using HDF5
  0.784713 seconds (407.98 k allocations: 27.318 MiB, 70.97% compilation time)

(HDF5TempEnv) pkg> st HDF5
      Status `C:\Users\kittisopikulm\HDF5TempEnv\Project.toml`
  [f67ccb44] HDF5 v0.16.0 `C:\Users\kittisopikulm\.julia\dev\HDF5
```

Reply:
That's quite a hit, perhaps we could 'hot-load' the filters as needed?

Reply:
If the overhead is Julia compilation (as opposed to HDF5 function calls), adding `precompile` statements for all the `@cfunction`s might improve things?

Reply:
I did use SnoopCompile and added the precompilation statements it recommended.

I suspect the main contribution to the added time comes from the additional dependencies. We've added CodecBzip2, CodecLz4, and CodecZstd. We then also need the load the corresponding JLL files.

Reply:
I can try to arrange this to only load the plugins if the underlying Codec* packages are loaded via Requires.jl. In this case, we may want to consider splitting the Blosc plugin out as well in the same manner.

Also note that this was all done on Windows, where I note package loading in Julia is slower for various reasons, including file system I/O speed.

Reply:
I switched to lazy loading all of the filters, **including Blosc**. Time to import HDF5 itself is reduced to 0.38 seconds.

I'm not really sure what tricks Requires.jl plays and how that interacts with precompilation.

```julia
julia> @time using HDF5
  0.380441 seconds (134.27 k allocations: 9.477 MiB, 36.10% compilation time)

julia> HDF5.Filters.FILTERS
Dict{Int32, Type{var"#s250"} where var"#s250"<:HDF5.Filters.Filter} with 6 entries:
  5 => NBit
  4 => Szip
  6 => ScaleOffset
  2 => Shuffle
  3 => Fletcher32
  1 => Deflate

julia> @time HDF5.Filters.register_filters()
  1.178848 seconds (2.32 M allocations: 140.625 MiB, 1.93% gc time, 74.63% compilation time)

julia> HDF5.Filters.FILTERS
Dict{Int32, Type{var"#s250"} where var"#s250"<:HDF5.Filters.Filter} with 10 entries:
  5     => NBit
  4     => Szip
  6     => ScaleOffset
  32015 => ZstdFilter
  2     => Shuffle
  307   => Bzip2Filter
  32004 => Lz4Filter
  3     => Fletcher32
  1     => Deflate
  32001 => BloscFilter
```

The filters can be loaded individually as needed.

```julia
julia> @time using HDF5
  0.369880 seconds (133.60 k allocations: 9.421 MiB, 35.51% compilation time)

julia> @time using Blosc
  0.434506 seconds (535.24 k allocations: 32.671 MiB, 75.02% compilation time)

julia> HDF5.Filters.FILTERS
Dict{Int32, Type{var"#s250"} where var"#s250"<:HDF5.Filters.Filter} with 7 entries:
  5     => NBit
  4     => Szip
  6     => ScaleOffset
  2     => Shuffle
  3     => Fletcher32
  1     => Deflate
  32001 => BloscFilter

julia> @time using CodecBzip2
  0.282867 seconds (820.43 k allocations: 49.120 MiB, 3.26% gc time, 70.65% compilation time)

julia> HDF5.Filters.FILTERS
Dict{Int32, Type{var"#s250"} where var"#s250"<:HDF5.Filters.Filter} with 8 entries:
  5     => NBit
  4     => Szip
  6     => ScaleOffset
  2     => Shuffle
  307   => Bzip2Filter
  3     => Fletcher32
  1     => Deflate
  32001 => BloscFilter

julia> @time using CodecLz4
  0.354482 seconds (645.27 k allocations: 39.529 MiB, 2.06% gc time, 86.55% compilation time)

julia> HDF5.Filters.FILTERS
Dict{Int32, Type{var"#s250"} where var"#s250"<:HDF5.Filters.Filter} with 9 entries:
  5     => NBit
  4     => Szip
  6     => ScaleOffset
  2     => Shuffle
  307   => Bzip2Filter
  32004 => Lz4Filter
  3     => Fletcher32
  1     => Deflate
  32001 => BloscFilter

julia> @time using CodecZstd
  0.156603 seconds (320.64 k allocations: 19.679 MiB, 5.24% gc time, 55.59% compilation time)

julia> HDF5.Filters.FILTERS
Dict{Int32, Type{var"#s250"} where var"#s250"<:HDF5.Filters.Filter} with 10 entries:
  5     => NBit
  4     => Szip
  6     => ScaleOffset
  32015 => ZstdFilter
  2     => Shuffle
  307   => Bzip2Filter
  32004 => Lz4Filter
  3     => Fletcher32
  1     => Deflate
  32001 => BloscFilter
```